File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeCategoryAdapter.java
package com.veganbeauty.app.features.shop.home;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;
import com.veganbeauty.app.R;
import com.veganbeauty.app.databinding.ShopHomeCategoryBinding;
import com.veganbeauty.app.features.shop.home.models.CategoryUiModel;

public class ShopHomeCategoryAdapter extends ListAdapter<CategoryUiModel, ShopHomeCategoryAdapter.CategoryViewHolder> {

    public interface OnCategoryClickListener {
        void onCategoryClick(CategoryUiModel category);
    }

    private final OnCategoryClickListener onCategoryClick;

    public ShopHomeCategoryAdapter(OnCategoryClickListener onCategoryClick) {
        super(new DiffCallback());
        this.onCategoryClick = onCategoryClick;
    }

    @NonNull
    @Override
    public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ShopHomeCategoryBinding binding = ShopHomeCategoryBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false
        );
        return new CategoryViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class CategoryViewHolder extends RecyclerView.ViewHolder {
        private final ShopHomeCategoryBinding binding;

        public CategoryViewHolder(ShopHomeCategoryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        @SuppressLint("ClickableViewAccessibility")
        public void bind(CategoryUiModel category) {
            binding.tvCategoryName.setText(category.getName());
            binding.tvProductCount.setText(category.getProductCount() + " sản phẩm");
            binding.ivCategoryIcon.setImageResource(category.getIconRes());

            binding.getRoot().setOnClickListener(v -> {
                if (onCategoryClick != null) onCategoryClick.onCategoryClick(category);
            });

            float density = binding.getRoot().getContext().getResources().getDisplayMetrics().density;
            binding.getRoot().setOnTouchListener((view, event) -> {
                MaterialCardView card = (MaterialCardView) view;
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        card.animate()
                                .scaleX(1.05f)
                                .scaleY(1.05f)
                                .translationZ(6f * density)
                                .setDuration(120)
                                .start();
                        card.setStrokeColor(ContextCompat.getColor(card.getContext(), R.color.primary));
                        card.setStrokeWidth((int) (2.5f * density));
                        break;
                    case MotionEvent.ACTION_UP:
                        card.animate()
                                .scaleX(1.0f)
                                .scaleY(1.0f)
                                .translationZ(0f)
                                .setDuration(150)
                                .start();
                        card.setStrokeColor(Color.parseColor("#EAEAEA"));
                        card.setStrokeWidth((int) (1f * density));
                        card.performClick();
                        break;
                    case MotionEvent.ACTION_CANCEL:
                        card.animate()
                                .scaleX(1.0f)
                                .scaleY(1.0f)
                                .translationZ(0f)
                                .setDuration(150)
                                .start();
                        card.setStrokeColor(Color.parseColor("#EAEAEA"));
                        card.setStrokeWidth((int) (1f * density));
                        break;
                }
                return true;
            });
        }
    }

    public static class DiffCallback extends DiffUtil.ItemCallback<CategoryUiModel> {
        @Override
        public boolean areItemsTheSame(@NonNull CategoryUiModel oldItem, @NonNull CategoryUiModel newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull CategoryUiModel oldItem, @NonNull CategoryUiModel newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeCategoryAdapter.kt
package com.veganbeauty.app.features.shop.home

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.veganbeauty.app.databinding.ShopHomeCategoryBinding
import com.veganbeauty.app.features.shop.home.models.CategoryUiModel

class ShopHomeCategoryAdapter(
    private val onCategoryClick: (CategoryUiModel) -> Unit
) : ListAdapter<CategoryUiModel, ShopHomeCategoryAdapter.CategoryViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        val binding = ShopHomeCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CategoryViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class CategoryViewHolder(private val binding: ShopHomeCategoryBinding) : RecyclerView.ViewHolder(binding.root) {
        @android.annotation.SuppressLint("ClickableViewAccessibility")
        fun bind(category: CategoryUiModel) {
            binding.tvCategoryName.text = category.name
            binding.tvProductCount.text = "${category.productCount} sản phẩm"
            binding.ivCategoryIcon.setImageResource(category.iconRes)
            
            binding.root.setOnClickListener {
                onCategoryClick(category)
            }

            val density = binding.root.context.resources.displayMetrics.density
            binding.root.setOnTouchListener { view, event ->
                val card = view as com.google.android.material.card.MaterialCardView
                when (event.action) {
                    android.view.MotionEvent.ACTION_DOWN -> {
                        card.animate()
                            .scaleX(1.05f)
                            .scaleY(1.05f)
                            .translationZ(6f * density)
                            .setDuration(120)
                            .start()
                        card.strokeColor = androidx.core.content.ContextCompat.getColor(card.context, com.veganbeauty.app.R.color.primary)
                        card.strokeWidth = (2.5f * density).toInt()
                    }
                    android.view.MotionEvent.ACTION_UP -> {
                        card.animate()
                            .scaleX(1.0f)
                            .scaleY(1.0f)
                            .translationZ(0f)
                            .setDuration(150)
                            .start()
                        card.strokeColor = android.graphics.Color.parseColor("#EAEAEA")
                        card.strokeWidth = (1f * density).toInt()
                        card.performClick()
                    }
                    android.view.MotionEvent.ACTION_CANCEL -> {
                        card.animate()
                            .scaleX(1.0f)
                            .scaleY(1.0f)
                            .translationZ(0f)
                            .setDuration(150)
                            .start()
                        card.strokeColor = android.graphics.Color.parseColor("#EAEAEA")
                        card.strokeWidth = (1f * density).toInt()
                    }
                }
                true
            }
        }
    }

    class DiffCallback : DiffUtil.ItemCallback<CategoryUiModel>() {
        override fun areItemsTheSame(oldItem: CategoryUiModel, newItem: CategoryUiModel) = oldItem.id == newItem.id
        override fun areContentsTheSame(oldItem: CategoryUiModel, newItem: CategoryUiModel) = oldItem == newItem
    }
}
